seed/seed20.go

```diff
@@ -1493,20 +1483,49 @@ func (s *seed20) ModeSnaps(mode string) ([]*Snap, error) {
 	if s.mode != AllModes && mode != s.mode {
 		return nil, fmt.Errorf("metadata was loaded only for snaps for mode %s not %s", s.mode, mode)
 	}
-	snaps := s.snaps[s.essentialSnapsNum:]
-	modes := s.modes[s.essentialSnapsNum:]
-	nGuess := len(snaps)
-	ephemeral := mode != "run"
-	if ephemeral {
-		nGuess /= 2
-	}
-	res := make([]*Snap, 0, nGuess)
-	for i, snap := range snaps {
-		if snapModesInclude(modes[i], mode) {
-			res = append(res, snap)
+
+	snapsWithMode := make([]*Snap, 0)
+	for _, sn := range s.snaps[s.essentialSnapsNum:] {
+		// since we're handing out pointers here (and we might need to modify
+		// the slice of components it holds), we need to make a copy of the snap
+		copied := *sn
+		copied.Components = append([]Component(nil), sn.Components...)
+
+		ms, ok := s.modelSnaps[sn.SnapName()]
+		if !ok {
+			// snaps not in the model will be considered as run mode, and so
+			// will all of its components
+			if mode == "run" {
+				snapsWithMode = append(snapsWithMode, &copied)
+			}
+			continue
+		}
+
+		copied.Components = nil
```

Comment:
shouldn't setting copied.Components be done in the two paths of the if instead of above to then in one case throw away the copy?

Reply:
Thanks, done in 6cde692a2d6e2c4da03d8a4eeab5284fd3a9ee91.